### PR TITLE
Add support for additional OAuth attributes in DCR

### PR DESCRIPTION
**Pull request by SachiniSiriwardene**
### Proposed changes in this pull request

Add support for storing, retrieving and updating additional OIDC attributes via DCR API.

- token_endpoint_auth_method
- token_endpoint_auth_signing_alg
- sector_identifier_uri
- id_token_signed_response_alg
- id_token_encrypted_response_alg
- id_token_encrypted_response_enc
- request_object_signing_alg
- software_statement
- tls_client_auth_subject_dn
- require_signed_request_object
- require_pushed_authorization_requests
- tls_client_certificate_bound_access_tokens
- subject_type
- request_object_encryption_alg
- request_object_encryption_enc

https://github.com/wso2/product-is/issues/16828 

Integration test PR - https://github.com/wso2/product-is/pull/16953

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java [hunk 1]
```diff
@@ -669,6 +683,75 @@ private void addOrUpdateOIDCSpProperty(OAuthAppDO oauthAppDO,
                 TOKEN_BINDING_VALIDATION, String.valueOf(oauthAppDO.isTokenBindingValidationEnabled()),
                 prepStatementForPropertyAdd, preparedStatementForPropertyUpdate);
 
+        addOrUpdateOIDCSpProperty(preprocessedClientId, spTenantId, spOIDCProperties,
+                TOKEN_AUTH_METHOD, String.valueOf(oauthAppDO.getTokenEndpointAuthMethod()),
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java:687]: As these getters used are returning String values, cant we straight away get the value from the getters without passing it to String.valueOf() method. Please check the other places as well in the PR.

File: components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java [hunk 2]
```diff
@@ -1,11 +1,11 @@
 package org.wso2.carbon.identity.oauth2.dcr.endpoint.dto;
 
-import com.fasterxml.jackson.annotation.JsonProperty;
-import io.swagger.annotations.ApiModel;
-import io.swagger.annotations.ApiModelProperty;
-
 import java.util.ArrayList;
 import java.util.List;
+
+import io.swagger.annotations.*;
```
Comment (ChinthakaJ98) [inline components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java:6]: Better to add the imports separately. Please check the other places too.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java (hunk 1 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dao/OAuthAppDAO.java]: fixed with 09b42394545a8f7aad2e358fe50edf8e28e75047

File: components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java (hunk 2 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java]: fixed with d2caa0cf1e61446265197460eab21fc1bd921988

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java [hunk 3]
```diff
@@ -33,7 +33,24 @@ public class Application implements Serializable {
...
+    private String requestObjectSignatureAlgorithm = null;
+    private String tlsClientAuthSubjectDN = null;
+    private boolean requirePushedAuthorizationRequests;
+    private boolean tlsClientCertificateBoundAccessTokens;
+    private String subjectType = null;
+    private String requestObjectEncryptionAlgorithm = null;
+    private String requestObjectEncryptionMethod = null;
+    private boolean isRequestObjectSignatureValidationEnabled;
+    private String idTokenEncryptionAlgorithm = null;
+    private String idTokenEncryptionMethod = null;
```
[the remark this answers is not in the document]
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java:51]: fixed with d2caa0cf1e61446265197460eab21fc1bd921988

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java [hunk 4]
```diff
@@ -83,7 +100,149 @@ public void setGrantTypes(List<String> grantTypes) {
 
         this.grantTypes = grantTypes;
     }
+    public String getJwksURI() {
+        return jwksURI;
```
[the remark this answers is not in the document]
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java:104]: fixed with d2caa0cf1e61446265197460eab21fc1bd921988

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java [hunk 5]
```diff
@@ -83,7 +100,149 @@ public void setGrantTypes(List<String> grantTypes) {
...
+        this.idTokenEncryptionAlgorithm = idTokenEncryptionAlgorithm;
+    }
+
+    public String getIdTokenEncryptionMethod() {
+        return idTokenEncryptionMethod;
+    }
+
+    public void setIdTokenEncryptionMethod(String idTokenEncryptionMethod) {
+        this.idTokenEncryptionMethod = idTokenEncryptionMethod;
+    }
```
[the remark this answers is not in the document]
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/bean/Application.java:260]: fixed with d2caa0cf1e61446265197460eab21fc1bd921988

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 6]
```diff
@@ -267,8 +331,10 @@ public Application updateApplication(ApplicationUpdateRequest updateRequest, Str
             throw DCRMUtils.generateServerException(
                     DCRMConstants.ErrorMessages.FAILED_TO_UPDATE_APPLICATION, clientId, e);
         }
-
-        return buildResponse(getApplicationById(clientId));
+        OAuthConsumerAppDTO oAuthConsumerAppDTO = getApplicationById(clientId);
+        //setting the jwksURI to be sent in the response
```
[the remark this answers is not in the document]
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java]: fixed with d2caa0cf1e61446265197460eab21fc1bd921988

File: components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java [hunk 7]
```diff
@@ -514,6 +640,14 @@ private ServiceProvider createServiceProvider(String applicationOwner, String te
         sp.setDescription("Service Provider for application " + spName);
         sp.setManagementApp(isManagementApp);
 
+        if (OAuthServerConfiguration.getInstance().isFapiSecurity()) {
+            // Add FAPI conformant application nad isThirdParty property to the service provider.
+            Map<String, Object> spProperties = new HashMap<>();
+            spProperties.put(OAuthConstants.IS_FAPI_CONFORMANT_APP, true);
+            spProperties.put(OAuthConstants.IS_THIRD_PARTY_APP, true);
```
Comment (SachiniSiriwardene) [inline components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:647]: add third party property irrespective of enabling or not

Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.dcr/src/main/java/org/wso2/carbon/identity/oauth/dcr/service/DCRMService.java:647]: add separate property for enabling/disabling FAPI validations in DCR

**Comment by SachiniSiriwardene** [comment]
Asgardeo issue - https://github.com/wso2-enterprise/asgardeo-product/issues/19895

File: components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/ApplicationDTO.java [hunk 8]
```diff
@@ -33,6 +33,24 @@ public class ApplicationDTO  {
   
   
   private String clientName = null;
+  private String jwksUri = null;
+  private String tokenEndpointAuthMethod = null;
+  private String tokenEndpointAuthSigningAlg = null;
+  private String sectorIdentifierUri = null;
+  private String idTokenSignedResponseAlg = null;
+  private String idTokenEncryptedResponseAlg = null;
+  private String idTokenEncryptedResponseEnc = null;
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/ApplicationDTO.java:42]: this variable name and `authorizationEncryptedResponseEnc` variable name is bit confusing as they are not self explantory due to having encryption twice in the name. this reads as  "idToken Encrypted Response Encryption"(not sure whether last enc was for encryption). Lets change both variable names to give more clarity.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java [hunk 9]
```diff
@@ -52,7 +53,25 @@ public class OAuthConsumerAppDTO {
...
+    private String requestObjectSignatureAlgorithm;
+    private String tlsClientAuthSubjectDN;
+    private boolean requirePushedAuthorizationRequests;
+    private boolean tlsClientCertificateBoundAccessTokens;
+    private String subjectType;
+    private String requestObjectEncryptionAlgorithm;
+    private String requestObjectEncryptionMethod;
+
+    public String getJwksURI() {
+        return jwksURI;
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java:70]: Add a new line after the method definition. Fix other similar places as well.

Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java:71]: fixed in f407f26a90e8117f69887dafa623d4f40a699ff3

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java [hunk 10]
```diff
@@ -300,5 +321,122 @@ public void setTokenBindingValidationEnabled(boolean tokenBindingValidationEnabl
 
         this.tokenBindingValidationEnabled = tokenBindingValidationEnabled;
     }
+
+    public String getTokenEndpointAuthMethod() {
+        return tokenEndpointAuthMethod;
+    }
+
+    public void setTokenEndpointAuthMethod(String tokenEndpointAuthMethod) {
+        this.tokenEndpointAuthMethod = tokenEndpointAuthMethod;
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java:330]: Add a new line after the method definition. Check other places as well.

Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java:332]: fixed

File: components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java [hunk 11]
```diff
@@ -40,14 +41,28 @@ public class RegistrationRequestDTO  {
...
   private boolean extPkceSupportPlain;
   private boolean extPublicClient;
-  
+  private String tokenEndpointAuthMethod = null;
+  private String tokenEndpointAuthSigningAlg = null;
+  private String sectorIdentifierUri = null;
+  private String idTokenSignedResponseAlg = null;
+  private String idTokenEncryptedResponseAlg = null;
+  private String idTokenEncryptedResponseEnc = null;
+  private String authorizationSignedResponseAlg = null;
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java:50]: With the last discussion, we can remove this right?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java [hunk 12]
```diff
@@ -52,7 +52,28 @@ public class OAuthConsumerAppDTO {
...
+    private String subjectType;
+    private String requestObjectEncryptionAlgorithm;
+    private String requestObjectEncryptionMethod;
+    private String jwksURI;
 
+    public String getJwksURI() {
+
+        return jwksURI;
+    }
+    public void setJwksURI(String jwksURi) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/dto/OAuthConsumerAppDTO.java:73]: Add a new line between methods

File: components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java (hunk 11 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.api.server.dcr/src/gen/java/org/wso2/carbon/identity/oauth2/dcr/endpoint/dto/RegistrationRequestDTO.java]: removed

**Review by janakamarasena — APPROVED** [review]

**Pull request merged**